File: foxy/parent/src/command/vanilla/entertainment/declarations/FunCommand.ts
import { ApplicationCommandOptionTypes, ApplicationCommandTypes } from "discordeno/types";
import { createCommand } from "../../../structures/createCommand";
import UnleashedCommandExecutor from "../../../structures/UnleashedCommandExecutor";
import StonksExecutor from "../memes/StonksExecutor";
import NotStonksExecutor from "../memes/NotStonksExecutor";
import LaranjoExecutor from "../memes/LaranjoExecutor";
import GostoExecutor from "../memes/GostoExecutor";
import GirlfriendExecutor from "../memes/GirlfriendExecutor";
import ErrorExecutor from "../memes/ErrorExecutor";
import ModaExecutor from "../memes/ModaExecutor";
import EminemExecutor from "../memes/8MileExecutor";

const FunCommand = createCommand({
    name: "fun",
    category: "fun",
    description: "[Memes] Create amazing memes, be it image or video",
    descriptionLocalizations: {
        "pt-BR": "[Memes] Crie memes incríveis, sejam eles imagem ou vídeo"
    },

    options: [{
        name: "eminem",
        description: "Generates an vídeo with a scene from the movie 8 Mile",
        descriptionLocalizations: {
            "pt-BR": "Gera um vídeo com uma cena do filme 8 Mile e um áudio customizado"
        },
        type: ApplicationCommandOptionTypes.SubCommandGroup,
        options: [{
            name: "8mile",
            description: "[Video] Generates an vídeo with a scene from the movie 8 Mile",
            descriptionLocalizations: {
                "pt-BR": "[Video] Gera um vídeo com uma cena do filme 8 Mile e um áudio customizado"
            },
            type: ApplicationCommandOptionTypes.SubCommand,
            options: [{
                name: "video_or_audio",
                description: "Video or Audio to be used in the video",
                descriptionLocalizations: {
                    "pt-BR": "Vídeo ou Áudio a ser usado no vídeo"
                },
                type: ApplicationCommandOptionTypes.Attachment,
                required: true
            }]
        }]
    },
    {
        name: "antes_que_vire_moda",
        description: "[Image] Brazilian meme: 'Before it becomes a trend, I'm warning you that I'm a fan of:'",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Montagem do meme 'Antes que vire moda, já vou avisando que sou fã de:'"
        },
        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "image",
            description: "Image to be used in the meme",
            descriptionLocalizations: {
                "pt-BR": "Imagem a ser usada no meme"
            },
            type: ApplicationCommandOptionTypes.Attachment,
            required: true
        }]
    },
    {
        name: "windowserror",
        description: "[Image] Generates a Windows error message",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Gera uma mensagem de erro do Windows"
        },
        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "text",
            description: "Message to be displayed in the error",
            descriptionLocalizations: {
                "pt-BR": "Mensagem a ser exibida no erro"
            },
            type: ApplicationCommandOptionTypes.String,
            required: true
        }]
    },
    {
        name: "girlfriend",
        description: "[Image] Who is your girlfriend?",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Quem é sua namorada?"
        },
        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "user",
            description: "User to be used in the meme",
            descriptionLocalizations: {
                "pt-BR": "Usuário a ser usado no meme"
            },
            type: ApplicationCommandOptionTypes.User,
            required: true
        }]
    },
    {
        name: "gosto",
        description: "[Image] Show a Brazilian meme called 'Não, não somos iguais'",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Mostre um meme brasileiro chamado 'Não, não somos iguais'"
        },
        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "asset1",
            description: "First asset to be used in the meme",
            descriptionLocalizations: {
                "pt-BR": "Primeiro asset a ser usado no meme"
            },
            type: ApplicationCommandOptionTypes.Attachment,
            required: true
        },
        {
            name: "asset2",
            description: "Second asset to be used in the meme",
            descriptionLocalizations: {
                "pt-BR": "Segundo asset a ser usado no meme"
            },
            type: ApplicationCommandOptionTypes.Attachment,
            required: true
        },
        {
            name: "text",
            description: "Text to be displayed in the meme",
            descriptionLocalizations: {
                "pt-BR": "Texto a ser exibido no meme"
            },
            type: ApplicationCommandOptionTypes.String,
            required: true
        }]
    },
    {
        name: "laranjo",
        description: "[Image] Create a Brazilian meme called 'Laranjo'",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Crie um meme brasileiro chamado 'Laranjo'"
        },

        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "text",
            description: "Text to be displayed in the meme",
            descriptionLocalizations: {
                "pt-BR": "Texto a ser exibido no meme"
            },
            type: ApplicationCommandOptionTypes.String,
            required: true
        }]
    },
    {
        name: "not_stonks",
        description: "[Image] Create a meme called 'Not Stonks'",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Crie um meme chamado 'Not Stonks'"
        },
        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "text",
            description: "Text to be displayed in the meme",
            descriptionLocalizations: {
                "pt-BR": "Texto a ser exibido no meme"
            },
            type: ApplicationCommandOptionTypes.String,
            required: true
        }]
    },
    {
        name: "stonks",
        description: "[Image] Create a meme called 'Stonks'",
        descriptionLocalizations: {
            "pt-BR": "[Imagem] Crie um meme chamado 'Stonks'"
        },
        type: ApplicationCommandOptionTypes.SubCommand,
        options: [{
            name: "text",
            description: "Text to be displayed in the meme",
            descriptionLocalizations: {
                "pt-BR": "Texto a ser exibido no meme"
            },
            type: ApplicationCommandOptionTypes.String,
            required: true
        }]
    }],
    type: ApplicationCommandTypes.ChatInput,

    execute: async (context: UnleashedCommandExecutor, endCommand, t) => {
        switch (context.getSubCommand()) {
            case "8mile":
                new EminemExecutor().execute(context, endCommand, t);
                break;
            case "antes_que_vire_moda":
                new ModaExecutor().execute(context, endCommand, t);
                break;
            case "windowserror":
                new ErrorExecutor().execute(context, endCommand, t);
                break;
            case "girlfriend":
                new GirlfriendExecutor().execute(context, endCommand, t);
                break;
            case "gosto":
                new GostoExecutor().execute(context, endCommand, t);
                break;
            case "laranjo":
                new LaranjoExecutor().execute(context, endCommand, t);
                break;
            case "not_stonks":
                new NotStonksExecutor().execute(context, endCommand, t);
                break;
            case "stonks":
                new StonksExecutor().execute(context, endCommand, t);
                break;
        }
    }
});

export default FunCommand;

File: foxy/src/main/kotlin/net/cakeyfox/foxy/command/vanilla/entertainment/declarations/FunCommand.kt
package net.cakeyfox.foxy.command.vanilla.entertainment.declarations

import net.cakeyfox.foxy.command.structure.FoxySlashCommandDeclarationWrapper
import net.cakeyfox.foxy.command.vanilla.entertainment.media.*
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData

class FunCommand : FoxySlashCommandDeclarationWrapper {
    override fun create() = command(
        "fun",
        "fun.description"
    ) {
        subCommandGroup(
            "eminem",
            "fun.eminem.description",
            baseName = this@command.name
        ) {
            subCommand(
                "8mile",
                "fun.eminem.8mile.description",
                block = {
                    executor = EminemExecutor()
                    addOption(
                        OptionData(
                            OptionType.ATTACHMENT,
                            "video_or_audio",
                            "fun.eminem.8mile.option.description",
                            true
                        ),
                        isSubCommand = true,
                        baseName = this@command.name
                    )
                }
            )
        }

        subCommand(
            "antes_que_vire_moda",
            "fun.moda.description",
            block = {
                executor = AntesQueVireModaExecutor()
                addOption(
                    OptionData(
                        OptionType.ATTACHMENT,
                        "image",
                        "fun.moda.option.description",
                        true
                    ),
                    isSubCommand = true,
                    baseName = this@command.name
                )
            }
        )

        subCommand(
            "windows_error",
            "fun.windows_error.description"
        ) {
            executor = ErrorExecutor()
            addOption(
                OptionData(
                    OptionType.STRING,
                    "message",
                    "fun.windows_error.option.description",
                    true
                ),
                isSubCommand = true,
                baseName = this@command.name
            )
        }

        subCommand(
            "girlfriend",
            "fun.girlfriend.description"
        ) {
            executor = GirlfriendMemeExecutor()
            addOption(
                OptionData(
                    OptionType.USER,
                    "user",
                    "fun.girlfriend.option.description",
                    true
                ),
                isSubCommand = true,
                baseName = this@command.name
            )
        }

        subCommand(
            "nao_somos_iguais",
            "fun.nao_somos_iguais.description"
        ) {
            executor = GostosIguaisExecutor()
            baseName = this@command.name

            addOptions(
                listOf(
                    OptionData(
                        OptionType.ATTACHMENT,
                        "first_image",
                        "fun.nao_somos_iguais.option.first_image",
                        true
                    ),

                    OptionData(
                        OptionType.ATTACHMENT,
                        "second_image",
                        "fun.nao_somos_iguais.option.second_image",
                        true
                    ),

                    OptionData(
                        OptionType.STRING,
                        "text",
                        "fun.nao_somos_iguais.option.text",
                        false
                    )
                ),
                isSubCommand = true,
                baseName = this@command.name,
            )
        }

        subCommand(
            "laranjo",
            "fun.laranjo.description"
        ) {
            executor = LaranjoExecutor()
            addOption(
                OptionData(
                    OptionType.STRING,
                    "text",
                    "fun.laranjo.option.description",
                    true
                ),
                isSubCommand = true,
                baseName = this@command.name
            )
        }

        subCommand(
            "stonks",
            "fun.stonks.description"
        ) {
            executor = StonksExecutor()
            addOption(
                OptionData(
                    OptionType.STRING,
                    "text",
                    "fun.stonks.option.description",
                    true
                ),
                isSubCommand = true,
                baseName = this@command.name
            )
        }

        subCommand(
            "not_stonks",
            "fun.not_stonks.description"
        ) {
            executor = NotStonksExecutor()
            addOption(
                OptionData(
                    OptionType.STRING,
                    "text",
                    "fun.not_stonks.option.description",
                    true
                ),
                isSubCommand = true,
                baseName = this@command.name
            )
        }
    }
}